examples/halo_finder/ArborX_HaloFinder.hpp

```diff
@@ -0,0 +1,476 @@
+      }
+    }
+    return curr;
+  }
+
+  template <typename Query, typename Insert>
+  KOKKOS_FUNCTION void operator()(Query const &query, int j,
+                                  Insert const &) const
+  {
+    int const i = ArborX::getData(query);
+
+    // only process edge in one direction
+    if (i > j)
+    {
+      // initialize to the first neighbor that's smaller
+      if (Kokkos::atomic_compare_exchange(&stat_(i), i, j) == i)
+        return;
+
+      {
+        // Per [1]:
+        //
+        // ```
+        // ...checks if the representatives of the two endpoints of the edge are
+        // the same. If so, nothing needs to be done. If not, the parent of the
+        // larger of the two representatives is updated to point to the smaller
+        // representative using an atomic CAS operation (<snip> depending on
+        // which representative is smaller). The atomic CAS is required because
+        // the representative may have been changed by another thread between
+        // the call to `representative()` and the call to the CAS. If it has
+        // been changed, `ostat` or `vstat` is updated with the latest value and
+        // the do-while loop repeats the computation until it succeeds, i.e.,
+        // until there is no data race on the parent.
+        // ```
+
+        // ##### ECL license (see LICENSE.ECL) #####
+        int vstat = representative(i);
+        int ostat = representative(j);
+
+        bool repeat;
+        do
+        {
+          repeat = false;
+          if (vstat != ostat)
+          {
+            int ret;
+            if (vstat < ostat)
+            {
+              if ((ret = Kokkos::atomic_compare_exchange(&stat_(ostat), ostat,
+                                                         vstat)) != ostat)
+              {
+                ostat = ret;
+                repeat = true;
+              }
+            }
+            else
+            {
+              if ((ret = Kokkos::atomic_compare_exchange(&stat_(vstat), vstat,
+                                                         ostat)) != vstat)
+              {
+                vstat = ret;
+                repeat = true;
+              }
+            }
+          }
+        } while (repeat);
+      }
+    }
+  }
+};
+
+template <typename ExecutionSpace, typename Primitives,
+          typename HalosIndicesView, typename HalosOffsetView>
+void findHalos(ExecutionSpace exec_space, Primitives const &primitives,
+               HalosIndicesView &halos_indices, HalosOffsetView &halos_offset,
+               float linking_length, int min_size = 2, bool verbose = false,
+               bool verify = false)
+{
+  static_assert(Kokkos::is_view<HalosIndicesView>{}, "");
+  static_assert(Kokkos::is_view<HalosOffsetView>{}, "");
+  static_assert(std::is_same<typename HalosIndicesView::value_type, int>::value,
+                "");
+  static_assert(std::is_same<typename HalosOffsetView::value_type, int>::value,
+                "");
```

Comment:
Consistency: either rely on the implicit conversion to bool (my preference) or use the `value` member type but don't mix them.